src/server/internalServer.cpp

```diff
@@ -95,35 +95,200 @@ inline std::string normalizeRootUrl(std::string rootUrl)
+}
+
+ParameterizedMessage rawEntryNotFoundMsg(const std::string& dt, const std::string& entry)
+{
+  return ParameterizedMessage("raw-entry-not-found",
+                              {
+                                {"DATATYPE", dt},
+                                {"ENTRY", entry},
+                              }
+  );
+}
+
+ParameterizedMessage tooManyBooksMsg(size_t nbBooks, size_t limit)
+{
+  return ParameterizedMessage("too-many-books",
+                              {
+                                {"NB_BOOKS", nbBooks},
+                                {"LIMIT", limit},
+                              }
+  );
+}
+
+ParameterizedMessage nonParameterizedMessage(const std::string& msgId)
+{
+  const ParameterizedMessage::Parameters noParams;
+  return ParameterizedMessage(msgId, noParams);
+}
+
+struct Error : public std::runtime_error {
+  explicit Error(const ParameterizedMessage& message)
+    : std::runtime_error("Error while handling request"),
+      _message(message)
+  {}
+
+  const ParameterizedMessage& message() const
+  {
+    return _message;
+  }
+
+  const ParameterizedMessage _message;
+};
+
+void checkBookNumber(const Library::BookIdSet& bookIds, size_t limit) {
+  if (bookIds.empty()) {
+    throw Error(nonParameterizedMessage("no-book-found"));
+  }
+  if (limit > 0 && bookIds.size() > limit) {
+    throw Error(tooManyBooksMsg(bookIds.size(), limit));
+  }
+}
+
 } // unnamed namespace
 
-SearchInfo::SearchInfo(const std::string& pattern)
-  : pattern(pattern),
-    geoQuery()
-{}
+std::pair<std::string, Library::BookIdSet> InternalServer::selectBooks(const RequestContext& request, const std::string& prefix) const
+{
+  // Try old API
+  try {
+    auto bookName = request.get_argument("content");
+    try {
+      const auto bookIds = Library::BookIdSet{mp_nameMapper->getIdForName(bookName)};
+      const auto queryString = request.get_query([&](const std::string& key){return key == "content";}, true);
+      return {queryString, bookIds};
+    } catch (const std::out_of_range&) {
+      throw Error(noSuchBookErrorMsg(bookName));
+    }
+  } catch(const std::out_of_range&) {
+    // We've catch the out_of_range of get_argument
+    // continue
+  }
 
-SearchInfo::SearchInfo(const std::string& pattern, GeoQuery geoQuery)
-  : pattern(pattern),
-    geoQuery(geoQuery)
-{}
+  // Does user directly gives us ids ?
+  try {
+    auto id_vec = request.get_arguments(prefix+"id");
+    if (id_vec.empty()) {
+      throw Error(noValueForArgMsg(prefix+"id"));
+    }
+    for(const auto& bookId: id_vec) {
+      try {
+        // This is a silly way to check that bookId exists
+        mp_nameMapper->getNameForId(bookId);
+      } catch (const std::out_of_range&) {
+        throw Error(noSuchBookErrorMsg(bookId));
+      }
+    }
+    const auto bookIds = Library::BookIdSet(id_vec.begin(), id_vec.end());
+    checkBookNumber(bookIds, m_multizimSearchLimit);
```

Comment:
1. `checkBookNumber()` is called at too many locations. It is better to have this check at a single location (right after the call to `selectBooks()`)
2. Please add a unit test covering the failure of this check